File: utils/per_class_stats.m
function [class_mu, class_sigma, class_cov] = per_class_stats(classes,w,h)
% measure the min and max bbox sizes for rejecting bad predictions
area = log(w.*h);
aspect_ratio = log(w./h);
uc = unique(classes(:)); 
n = numel(uc);

class_mu = zeros(n,4,'single'); % width, height, area, aspect_ratio
class_sigma = zeros(n,4,'single'); 
class_cov = zeros(n,4,4,'single');

for i = 1:n
    j = find(classes==uc(i));
    wj = log(w(j));  hj = log(h(j));  aj = area(j);  arj = aspect_ratio(j,:);
    data = [wj, hj, aj, arj];
    class_mu(i,:) = mean(data,1);
    class_sigma(i,:) = std(data,1);
    class_cov(i,:,:) = cov(data);
    %[~,C] = kmeans([wj hj],1,'MaxIter',5000,'OnlinePhase','on');
    
    %class_name = xview_names(xview_classes2indices(uc(i)));
    %close all; hist211(arj,hj,40); title(sprintf('%s, cor = %g',class_name,corr(wj,hj)))
    
    % close all; hist211(wj,hj,{linspace(0,max(wj),40),linspace(0,max(hj),40)}); plot(C(:,1),C(:,2),'g.','MarkerSize',50); title(sprintf('%s, cor = %g',class_name,corr(wj,hj)))
    % ha=fig; histogram(arj, linspace(-3,3,50)); title(class_name); ha.YScale='linear';
end
end

File: src/targets/per_class_stats.py
import numpy as np

def per_class_stats(classes,w,h):
    """
    Function to calculate statistics of target data.
    
    | **Inputs:**
    |    *classes:* target data processed/produced with the Target class
    |    *w:* image width
    |    *h:* image height

    | **Outputs:**
    |    *class_mu:* mean of target classes
    |    *class_sigma:* standard deviation of target classes
    |    *class_cov:* covariance of target classes
    """
    area         = np.log(w*h)
    aspect_ratio = np.log(w/h)
    uc           = np.unique(classes)
    n            = np.size(uc)
    class_mu     = np.zeros([n,4])
    class_sigma  = np.zeros([n,4])
    class_cov    = np.zeros([n,4,4])
    for i in range(n):
        j    = np.where(classes == uc[i])[0]
        wj   = np.log(w[j])
        hj   = np.log(h[j])
        aj   = area[j]
        arj  = aspect_ratio[j]
        data = np.vstack([wj,hj,aj,arj]).T
        class_mu[i]    = np.mean(data,axis=0)
        class_sigma[i] = np.std(data,axis=0)
        class_cov[i]   = np.cov(data.T)
    return class_mu, class_sigma, class_cov
